pipelineBackends/model_ort.go

```diff
@@ -62,62 +117,161 @@ func loadInputOutputMetaORT(onnxBytes []byte) ([]InputOutputInfo, []InputOutputI
+	total := N * L
+
+	// 2) filter out cache inputs and detect whether to append cache later
+	metas := make([]InputOutputInfo, 0, len(model.InputsMeta))
+	hasCache := false
+	for _, m := range model.InputsMeta {
+		if strings.HasPrefix(m.Name, "past_key") {
+			hasCache = true
+		} else {
+			metas = append(metas, m)
+		}
+	}
+
+	// 3) prepare result containers
+	inputVals := make([]ort.Value, len(metas))
+	masks := make([][]bool, N)
+
+	// 4) build each tensor
+	for mi, meta := range metas {
+		backing := make([]int64, total)
+		idx := 0
+
+		for bi, inp := range batch.Input {
+			seqLen := len(inp.TokenIDs)
+			padLen := L - seqLen
+			maskRow := make([]bool, L)
+
+			for pos := range L {
+				var v int64
+				switch meta.Name {
+				case "input_ids":
+					if padLeft {
+						if pos < padLen {
+							v = 0
+						} else {
+							v = int64(inp.TokenIDs[pos-padLen])
+							maskRow[pos] = true
+						}
+					} else {
+						if pos < seqLen {
+							v = int64(inp.TokenIDs[pos])
+							maskRow[pos] = true
+						}
 					}
-				} else {
-					backingSlice[counter] = 0 // pad with zero
+				case "token_type_ids":
+					// always right-pad
+					if pos < seqLen {
+						v = int64(inp.TypeIDs[pos])
+					}
+				case "attention_mask":
+					if padLeft {
+						if pos >= padLen {
+							v = int64(inp.AttentionMask[pos-padLen])
+						}
+					} else {
+						if pos < seqLen {
+							v = int64(inp.AttentionMask[pos])
+						}
+					}
+				case "position_ids":
+					// 1-indexed positions
+					v = int64(pos + 1)
+				default:
+					return fmt.Errorf("unrecognized input %q", meta.Name)
 				}
-				counter++
+
+				backing[idx] = v
+				idx++
 			}
 
-			if inputMeta.Name == "input_ids" {
-				paddingMasks[j] = inputPaddingMask
+			if meta.Name == "input_ids" {
+				masks[bi] = maskRow
 			}
 		}
-		inputTensors[i], tensorCreationErr = ort.NewTensor(ort.NewShape(int64(batchSize), int64(batch.MaxSequenceLength)), backingSlice)
-		if tensorCreationErr != nil {
-			return tensorCreationErr
+
+		// create the ONNX Runtime tensor
+		t, err := ort.NewTensor(ort.NewShape(int64(N), int64(L)), backing)
+		if err != nil {
+			return err
 		}
+		inputVals[mi] = t
 	}
-	batch.InputValues = inputTensors
-	batch.PaddingMask = paddingMasks
-	batch.DestroyInputs = func() error {
-		var destroyError error
-		for _, ortTensor := range inputTensors {
-			destroyError = errors.Join(destroyError, ortTensor.Destroy())
+
+	// 5) append KV‐cache tensors if needed
+	if hasCache {
+		cacheTensors, err := CreateCacheORT(
+			batch.Size,
+			model.NumHiddenLayers,
+			model.NumKeyValueHeads,
+			model.FixedCacheSize,
+			model.HeadDim,
+		)
+		if err != nil {
+			return err
 		}
-		return destroyError
+		inputVals = append(inputVals, cacheTensors...)
 	}
 
+	// 6) assign and prepare cleanup
+	batch.InputValues = inputVals
+	batch.PaddingMask = masks
+	batch.DestroyInputs = func() error {
+		var agg error
+		for _, t := range inputVals {
```

Comment:
should use batch.InputValues instead here, as it gets overwritten per generation, and we should destroy current